### PR TITLE
SEARCH-279 snapshot for inverted index

### PR DESCRIPTION
**Pull request by Dronplane**
### Scope & Purpose

PR adds snapshot usage for inverted indexes the same way as it  is already done for ArangoSearch view.
That is part of the fix for updates consistency.

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [x] Tests
  - [ ] **Regression tests**
  - [x] C++ **Unit tests**
  - [x] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Aql/IndexExecutor.cpp [hunk 1]
```diff
@@ -135,11 +146,17 @@ IndexIterator::CoveringCallback getCallback(
...
-    // move a document id
-    AqlValue v(AqlValueHintUInt(token.id()));
-    AqlValueGuard guard{v, true};
     TRI_ASSERT(!output.isFull());
-    output.moveValueInto(registerId, input, guard);
+    // move a document id
+    if constexpr (isLocalDocumentId) {
+      TRI_ASSERT(token.isSet());
+      AqlValue v(AqlValueHintUInt(token.id()));
+      AqlValueGuard guard{v, true};
```
Comment (gnusi) [inline arangod/Aql/IndexExecutor.cpp:154]: It seems that we can use `AqlValueGuard guard{v, true};` as AqlValue uses SSO

File: arangod/Aql/MaterializeExecutor.cpp [hunk 2]
```diff
@@ -129,71 +135,80 @@ void MaterializeExecutor<T>::fillBuffer(AqlItemBlockInputRange& inputRange) {
...
+        do {
+          ++doc;
+        } while (doc != end &&
+                 lastSegment == std::get<0>(_bufferedDocs[*doc]).segment());
+        continue;
       }
-      ++doc;
     }
+    TRI_ASSERT(docValue);
+    if (std::get<0>(document).doc() ==
```
Comment (gnusi) [inline arangod/Aql/MaterializeExecutor.cpp:184]: maybe 
```
if (const auto doc = std::get<0>(document).doc(); doc == pkReader->seek(doc)) {
```

File: arangod/IResearch/IResearchInvertedIndex.cpp [hunk 3]
```diff
@@ -526,17 +537,37 @@ class IResearchInvertedIndexIterator final
...
       } else {
         if constexpr (produce) {
-          if (_doc->value == _pkDocItr->seek(_doc->value)) {
-            LocalDocumentId documentId;
+          LocalDocumentId documentId;
+          // For !withCovering that means actual doc reading
+          // if combined with "produce".
+          // And we must read LocalDocumentId anyway.
+          // Otherwise we read it only if required.
+          constexpr bool needReadLocalDocumentId =
```
Comment (gnusi) [inline arangod/IResearch/IResearchInvertedIndex.cpp:545]: `static constexpr`

File: arangod/IResearch/IResearchInvertedIndex.cpp [hunk 4]
```diff
@@ -526,17 +537,37 @@ class IResearchInvertedIndexIterator final
...
             bool const readSuccess =
+                !needReadLocalDocumentId ||
                 DocumentPrimaryKey::read(documentId, _pkValue->value);
             if (readSuccess) {
               if constexpr (withCovering) {
                 _projections.seek(_doc->value);
-                if (callback(documentId, _projections)) {
+                TRI_ASSERT(_readerOffset > 0);
+                TRI_ASSERT(documentId.isSet() == emitLocalDocumentId);
+                bool emitRes{false};
```
Comment (gnusi) [inline arangod/IResearch/IResearchInvertedIndex.cpp:557]: bool emitRes;

File: arangod/IResearch/IResearchInvertedIndex.cpp [hunk 5]
```diff
@@ -624,24 +656,42 @@ class IResearchInvertedIndexMergeIterator final
...
-      auto& segment = _segments[_heap_it.value()];
+      auto const currentIdx = _heap_it.value();
+      auto& segment = _segments[currentIdx];
       if constexpr (produce) {
-        if (segment.doc->value == segment.pkDocItr->seek(segment.doc->value)) {
+        // For !withCovering that means actual doc reading
+        // if combined with "produce".
+        // And we must read LocalDocumentId anyway.
+        // Otherwise we read it only if required.
+        constexpr bool needReadLocalDocumentId =
```
Comment (gnusi) [inline arangod/IResearch/IResearchInvertedIndex.cpp:670]: static constexpr

File: arangod/IResearch/IResearchInvertedIndex.cpp [hunk 6]
```diff
@@ -624,24 +656,42 @@ class IResearchInvertedIndexMergeIterator final
...
-              if (callback(documentId, segment.projections)) {
+              SearchDoc doc(_snapshot.segment(currentIdx), segment.doc->value);
+              TRI_ASSERT(documentId.isSet() == emitLocalDocumentId);
+              bool emitRes{false};
+              if constexpr (emitLocalDocumentId) {
+                emitRes = callback(documentId, segment.projections);
+              } else {
+                emitRes = callback(aql::AqlValue{doc.encode(_buf)},
+                                   segment.projections);
+              }
```
Comment (gnusi) [inline arangod/IResearch/IResearchInvertedIndex.cpp:689]:
```suggestion
bool const emitRes = [&]() {
  if constexpr (emitLocalDocumentId) {
    return callback(documentId, segment.projections);  
   } else {
    return callback(aql::AqlValue{doc.encode(_buf)}, segment.projections);
  }
}();
```

File: arangod/IResearch/IResearchInvertedIndex.cpp [hunk 7]
```diff
@@ -953,23 +1013,40 @@ std::unique_ptr<IndexIterator> IResearchInvertedIndex::iteratorForCondition(
...
-          mutableConditionIdx);
+      if (opts.forLateMaterialization) {
+        return std::make_unique<IResearchInvertedIndexIterator<false>>(
+            monitor, collection, state, trx, node, &_meta, reference,
+            mutableConditionIdx);
+      } else {
+        return std::make_unique<IResearchInvertedIndexIterator<true>>(
+            monitor, collection, state, trx, node, &_meta, reference,
+            mutableConditionIdx);
+      }
```
Comment (gnusi) [inline arangod/IResearch/IResearchInvertedIndex.cpp:1024]:
```
auto resolveLateMaterialition = [&](auto&& factory) {
  if (opts.forLateMaterialization) {
      return factory(std::true_type{});
  } else {
      return factory(std::false_type{});
  }
};

if (_meta._sort.empty()) {
  return resolveLateMaterialition([&]<bool LateMaterialization>(std::integral_type<bool, LateMaterialization>) {
    return std::make_unique<IResearchInvertedIndexIterator<LateMaterialization>>(
            monitor, collection, state, trx, node, &_meta, reference,
            mutableConditionIdx);
  });
} else {
  return resolveLateMaterialition([&]<bool LateMaterialization>(std::integral_type<bool, LateMaterialization>) {
    return std::make_unique<IResearchInvertedIndexMergeIterator<LateMaterialization>>(
            monitor, collection, state, trx, node, &_meta, reference,
            mutableConditionIdx);
  });
}
```

File: arangod/IResearch/IResearchInvertedIndex.cpp [hunk 8]
```diff
@@ -920,31 +978,33 @@ std::unique_ptr<IndexIterator> IResearchInvertedIndex::iteratorForCondition(
...
+      auto selfLock = this->self()->lock();
+      if (!selfLock) {
+        THROW_ARANGO_EXCEPTION_MESSAGE(
+            TRI_ERROR_INTERNAL,
+            absl::StrCat("Failed to lock datastore for index '", index().name(),
+                         "'"));
       }
-
-      ctx->snapshot = snapshot();
+      // TODO(MBkkt) Move it to optimization stage
```
Comment (gnusi) [inline arangod/IResearch/IResearchInvertedIndex.cpp:996]: Move above lock acquisition

**Review by gnusi — APPROVED** [review]

**Comment by Dronplane** [comment]
Tests blue except known ASAN failures

**Pull request merged**